### PR TITLE
Path module Replacement Procedures

### LINKED ISSUE #16767: Should we add replacement functions to the Path module?
The Path module has functions to make basic low-level transformations on paths. However, there's no function for replacing parts of a path string, and the user would have to use string functions (slice, concat etc) to achieve that.

Python's `os.path` doesn't have such functions, but higher-level `pathlib` does have some methods to call on Path objects: https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.with_name

I think we should add some replacement functions to the Path module. I suggest the following:

- `replaceDirname(path, newDirname)` (or `replaceDirName`: see https://github.com/chapel-lang/chapel/issues/16764)
- `replaceBasename(path, newBasename)` (or `replaceBaseName`: see https://github.com/chapel-lang/chapel/issues/16764)
- `replaceExt(path, newExt)` (see https://github.com/chapel-lang/chapel/issues/16766)

One design question about them would be what happens if the path doesn't have dirname, basename, or extension. We can adopt pathlib's naming scheme, e.g. `withDirname`, `withBasename` etc, that would add the given part if it doesn't exist and replace if it does.

### PR DESCRIPTION
**Pull request by king-11**
Closes #16767 

This PR adds:

- `replaceDirname`
- `replaceBasename`
- `replaceExt`

to the Path module. These functions can be used to replace different parts
of a path string. The design for these functions is here:

  https://github.com/chapel-lang/chapel/issues/16767#issuecomment-806061865

Tests:
- [x] added tests pass with darwin, asan, memleaks
- [x] full linux64

### CODE REVIEW

**Review by e-kayrakli — CHANGES_REQUESTED** [review]
Thanks @king-11! I have some inline comments that are mostly stylistic. More important points:

- I think the basename argument can contain `/`
- This needs some correctness tests to avoid errors down the road. FWIW, I have written something like the following to test the correctness (and it looked good beside my first point, which was an intentional choice of yours that should change, I think)

```chapel
use Path;

// the original path doesn't have a basename. returns "/foo/bar/baz"
writeln(replaceBasename("/foo/bar/", "baz"));

// the original path doesn't have a dirname.  returns "bar/foo"
writeln(replaceDirname("foo", "bar"));

// the original path doesn't have an extension.  returns "foo.txt"
writeln(replaceExt("foo", ".txt"));


// preceding . in extension is optional
writeln(replaceExt("foo.bar", "txt"));
writeln(replaceExt("foo.bar", ".txt"));

try {
  writeln(replaceExt("foo/", "txt"));
}
catch e {
  writeln("Caught error: ", e);
}

writeln("replaceBasename tests");
writeln(replaceBasename("", ""));
writeln(replaceBasename("/", ""));
writeln(replaceBasename("foo/", ""));
writeln(replaceBasename("foo/bar", ""));
// writeln(replaceBasename("", "/"));  // Engin: I expect "" to be a basename,
                                       // and this to return "/"
writeln(replaceBasename("", "foo"));   // because this returns "foo", and I
                                       // think that's right
writeln(replaceBasename("foo/bar.txt", "baz"));

writeln("replaceDirname tests");
writeln(replaceDirname("", ""));
writeln(replaceDirname("/", ""));
writeln(replaceDirname("foo/", ""));
writeln(replaceDirname("foo/bar", ""));
writeln(replaceDirname("", "/"));  // Engin: I expect "" to be a basename,
                                       // and this to return "/"
writeln(replaceDirname("", "foo"));   // because this returns "foo", and I
                                       // think that's right
writeln(replaceDirname("/bar.txt", "baz"));
writeln(replaceDirname("/foo//./../bar.txt", "baz"));

writeln("replaceExt tests");
try {
  writeln(replaceExt("", ""));
}
catch e {
  writeln("Caught error :", e);
}
writeln(replaceExt(".bashrc", "newext"));
writeln(replaceExt(".exec.out.tmp", "newext"));
writeln(replaceExt("./foo", "newext"));
writeln(replaceExt("/./foo", "newext"));
try {
  writeln(replaceExt("/./", "newext"));
}
catch e {
  writeln("Caught error :", e);
}

```

This'll need to go into a `.chpl` file with an accompanying `.good` file that contains the expected output. Both of those file can go to `test/library/standard/Path/replace` (a new dir).

See more about our test system here: https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst

**Comment by king-11** [comment]
I have added some new tests besides the ones you provided and also handled an edge case that sneaked past where the extension provided is empty so made the necessary change for it as well.

File: modules/standard/Path.chpl [hunk 1]
```diff
@@ -810,6 +813,85 @@ proc file.relPath(start:string=curDir): string throws {
...
+*/
+proc replaceDirname(path: string, newDirname: string): string {
+    const (dirname, basename) = splitPath(path);
+    return joinPath(newDirname, basename);
+}
+
+/*
+  Returns a new path with basename replaced with the provided
+  new argument of basename. If path had no basename the argument is
+  added to the path or if the provided new basename is empty string then basename is removed from path.
```
Comment (e-kayrakli) [inline modules/standard/Path.chpl:836]: Could you make sure that your lines fit in 80 columns?

**Review by e-kayrakli — COMMENTED** [review]
I had only one minor style comment, otherwise I don't see anything wrong here. And thanks for adding a lot of tests on this!

@lydia-duncan -- could you also take a look at this? Unless you have major issues with it, I can run testing on it.

File: modules/standard/Path.chpl (hunk 1 above)
Comment (king-11) [reply modules/standard/Path.chpl]: corrected them to maintain 80 columns limit

**Comment by king-11** [comment]
@e-kayrakli did the test run fine 😃  or there were some issues I need to fix 😨  ?

**Comment by e-kayrakli** [comment]
Test status:

- [x] added tests with standard darwin, asan, memleaks
- [ ] full standard linux64

**Review by e-kayrakli — APPROVED** [review]

**Comment by e-kayrakli** [comment]
All clean!

I am planning to merge this on Monday.

**Comment by e-kayrakli** [comment]
@king-11 -- Thanks again for the contribution!

Note that I have added more description to your PR message. By convention, we copy/paste that PR message as the merge commit message so that it appears in the git history. Therefore, we want it to be a bit descriptive. For your next PRs it'd be great if you can describe what the PR does when you open it. See more here https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/ContributorInfo.rst#pull-request-description

**Comment by king-11** [comment]
@e-kayrakli I will take a look and make sure to write a descriptive PR Message from next time onwards.

**Pull request merged**